src/compositor/hwcimage.cpp

```diff
@@ -0,0 +1,324 @@
+{
+    static bool initialized = false;
+    if (initialized)
+        return;
+    initialized = true;
+
+    glEGLImageTargetTexture2DOES = (_glEGLImageTargetTexture2DOES) eglGetProcAddress("glEGLImageTargetTexture2DOES");
+    eglCreateImageKHR = (_eglCreateImageKHR) eglGetProcAddress("eglCreateImageKHR");
+    eglDestroyImageKHR = (_eglDestroyImageKHR) eglGetProcAddress("eglDestroyImageKHR");
+    eglHybrisCreateNativeBuffer = (_eglHybrisCreateNativeBuffer) eglGetProcAddress("eglHybrisCreateNativeBuffer");
+    eglHybrisLockNativeBuffer = (_eglHybrisLockNativeBuffer) eglGetProcAddress("eglHybrisLockNativeBuffer");
+    eglHybrisUnlockNativeBuffer = (_eglHybrisUnlockNativeBuffer) eglGetProcAddress("eglHybrisUnlockNativeBuffer");
+    eglHybrisReleaseNativeBuffer = (_eglHybrisReleaseNativeBuffer) eglGetProcAddress("eglHybrisReleaseNativeBuffer");
+}
+
+QSGTexture *HwcImageTexture::create(const QImage &image)
+{
+    hwcimage_initialize();
+
+    EGLClientBuffer buffer = 0;
+    int width = image.width();
+    int height = image.height();
+    int usage = HYBRIS_USAGE_SW_WRITE_RARELY | HYBRIS_USAGE_HW_TEXTURE | HYBRIS_USAGE_HW_COMPOSER;
+    int stride = 0;
+    int format = HYBRIS_PIXEL_FORMAT_RGBA_8888;
+    char *data;
+
+    eglHybrisCreateNativeBuffer(width, height, usage, format, &stride, &buffer);
+    Q_ASSERT(buffer);
+    eglHybrisLockNativeBuffer(buffer, HYBRIS_USAGE_SW_WRITE_RARELY, 0, 0, width, height, (void **) &data);
+    Q_ASSERT(data);
+
+    const int dbpl = stride * 4;
+    const int bpl = qMin(dbpl, image.bytesPerLine());
+    for (int y=0; y<height; ++y)
+        memcpy(data + y * dbpl, image.constScanLine(y), bpl);
+
+    eglHybrisUnlockNativeBuffer(buffer);
+
+    EGLImageKHR eglImage = eglCreateImageKHR(eglGetDisplay(EGL_DEFAULT_DISPLAY),
+                                             EGL_NO_CONTEXT,
+                                             EGL_NATIVE_BUFFER_HYBRIS,
+                                             buffer,
+                                             0);
+    Q_ASSERT(eglImage);
+
+    return new HwcImageTexture(buffer, eglImage, QSize(width, height));
+
+}
+
+
+HwcImageTexture::HwcImageTexture(EGLClientBuffer buffer, EGLImageKHR image, const QSize &size)
+    : m_buffer(buffer)
+    , m_image(image)
+    , m_size(size)
+    , m_bound(false)
+{
+    glGenTextures(1, &m_id);
+}
+
+HwcImageTexture::~HwcImageTexture()
+{
+    glDeleteTextures(1, &m_id);
+    eglDestroyImageKHR(eglGetDisplay(EGL_DEFAULT_DISPLAY), m_image);
+    eglHybrisReleaseNativeBuffer(m_buffer);
+}
+
+void HwcImageTexture::bind()
+{
+    updateBindOptions();
+    glBindTexture(GL_TEXTURE_2D, m_id);
+    if (!m_bound) {
+        m_bound = true;
+        glEGLImageTargetTexture2DOES(GL_TEXTURE_2D, m_image);
+    }
+}
+
+
+// Memory layout of ANativeWindowBuffer from Android's /system/window.h which
+// is the structure that libhybris uses internally to represent the actual
+// EGLClientBuffer. We need this to extract the gralloc handle.
```

Comment:
We can do that also, but this way we don't end up with a dependency on a given version of hybris (As long as its implementation doesn't change).